src/ray/raylet/worker_pool.cc

```diff
@@ -711,17 +714,28 @@ void WorkerPool::TryKillingIdleWorkers() {
       auto rpc_client = worker->rpc_client();
       RAY_CHECK(rpc_client);
       rpc::ExitRequest request;
-      rpc_client->Exit(request, [](const ray::Status &status, const rpc::ExitReply &r) {
-        if (!status.ok()) {
-          RAY_LOG(ERROR) << "Failed to send exit request: " << status.ToString();
-        }
-      });
-      // Remove the worker from the idle pool so it can't be popped anymore.
-      RemoveWorker(worker_state.idle, worker);
-      if (!worker->IsDead()) {
-        worker->MarkDead();
-        running_size--;
-      }
+      rpc_client->Exit(
+          request, [this, worker](const ray::Status &status, const rpc::ExitReply &r) {
+            if (!status.ok()) {
+              RAY_LOG(ERROR) << "Failed to send exit request: " << status.ToString();
+              return;
+            }
+
+            if (r.success()) {
+              auto &worker_state = GetStateForLanguage(worker->GetLanguage());
+              // If we could kill the worker properly, we remove them from the idle pool.
+              if (RemoveWorker(worker_state.idle, worker)) {
+                // If the worker is not idle at this moment, we don't mark them dead.
+                // It will be marked as dead after the current task is finished.
```

Comment:
> It will be marked as dead after the current task is finished.

Would you mind pointing out which part of the code ensures this?

Reply:
task_manager_->DrainAndShutdown() will exit the worker after the task is executed & `DisconnectWorker` in the worker pool will remove the worker from the idle pool